pkg/controller/common/manage_children.go

```diff
@@ -197,15 +207,98 @@ func deleteChildren(client *dynamicclientset.ResourceClient, parent *unstructure
 				}
 				continue
 			}
+
+			lastUpdateName := lastUpdateCacheKey(client, obj)
+			cacheLock.Lock()
+			delete(lastUpdatedCache, lastUpdateName)
+			cacheLock.Unlock()
 		}
 	}
 	return utilerrors.NewAggregate(errs)
 }
 
-func updateChildren(client *dynamicclientset.ResourceClient, updateStrategy ChildUpdateStrategy, parent *unstructured.Unstructured, observed, desired map[string]*unstructured.Unstructured) error {
+type lastUpdate struct {
+	hash               uint64
+	resourcegeneration int64
+}
+
+var (
+	lastUpdatedCache = make(map[string]*lastUpdate)
+	cacheLock        = &sync.RWMutex{}
+)
+
+func lastUpdateCacheKey(client *dynamicclientset.ResourceClient, obj *unstructured.Unstructured) string {
+	return fmt.Sprintf("%s/%s/%s/%s", client.Group, client.Kind, obj.GetNamespace(), obj.GetName())
+}
+
+func updateChildren(client *dynamicclientset.ResourceClient, updateStrategy ChildUpdateStrategy, parent *unstructured.Unstructured, observed, desired map[string]*unstructured.Unstructured, useServerSideApply bool) error {
 	var errs []error
+
 	for name, obj := range desired {
+		if useServerSideApply {
+			data, err := json.Marshal(obj)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+
+			hash := xxhash.Sum64(data)
+			lastUpdateCacheName := lastUpdateCacheKey(client, obj)
+			if oldObj := observed[name]; oldObj != nil {
+				cacheLock.RLock()
+				if lastUpdated, ok := lastUpdatedCache[lastUpdateCacheName]; ok {
+					cacheLock.RUnlock()
+					if lastUpdated.hash == hash && lastUpdated.resourcegeneration == oldObj.GetGeneration() {
+						logging.Logger.Info("Skipping update, no changes detected", "name", lastUpdateCacheName)
+						continue
+					}
+					logging.Logger.Info("Detected changes, updating", "name", lastUpdateCacheName)
+				} else {
+					cacheLock.RUnlock()
+					logging.Logger.Info("No cache entry found, updating", "name", lastUpdateCacheName)
+				}
+
+				//check if observed object hast last applied annotation
+				_, hasLastApplied := oldObj.GetAnnotations()[dynamicapply.LastAppliedAnnotation]
+				if hasLastApplied {
+					//if observed object has has last applied annotation we need to remove it
+					//from the remote object to avoid conflicts
+					nullifyLastAppliedAnnotation(oldObj)
+					_, err := client.Namespace(obj.GetNamespace()).Update(context.TODO(), oldObj, metav1.UpdateOptions{})
+					if err != nil {
+						logging.Logger.Error(err, "Failed to remove last applied annotation from observed object", "parent", parent, "child", obj)
+						errs = append(errs, err)
+						continue
+					}
+				}
+			}
+
+			patched, err := client.Namespace(obj.GetNamespace()).Patch(context.TODO(), obj.GetName(), types.ApplyPatchType, data, metav1.PatchOptions{
```

Comment:
I like how you used PatchOptions instead of ApplyOptions and making it SSA by specifying types.ApplyPatchType.

Reply:
Thank you!